### DISCUSSION TITLE
Do System.Linq.Expressions support 'IS DISTINCT FROM' and 'IS NOT DISTINCT FROM' with npgsql/postgres?

### DISCUSSION THREAD

**Discussion by cirrusone** [dotnet/runtime#102369] [Q&A] [answered]
Postgres supports statements `IS DISTINCT FROM` and `IS NOT DISTINCT FROM` which is a much nicer handling of null values, especially when inverting expressions using `NOT`.  [Pg wiki - IS DISTINCT FROM](https://wiki.postgresql.org/wiki/Is_distinct_from)

DISTINCT here is different to what '<>' or '!=" tests for.

Are these statements currently supported by System.Linq.Expressions?

If not, is it something that's likely to be supported in the near future? 

It would be a really useful feature to include in both `Npgsql` and `System.Linq.Expressions` and massively help with filter logic where currently applying NOT to an existing filter excludes null values and complicates building queries to handle where you do genuinely need to return nulls.

Edit: I didn't realise when I posted this that SQL Server 2022 now support these predicates too [IS [NOT] DISTINCT FROM (Transact-SQL)](https://learn.microsoft.com/en-us/sql/t-sql/queries/is-distinct-from-transact-sql?view=sql-server-ver16)

**Comment by huoyaoyuan** [comment, marked as answer]
It is not something for `System.Linq.Expressions`, but for providers like `Npgsql`.

Generally, for custom or extended functionalities, the LINQ-to-SQL provider can expose some custom functions. The functions are encoded as calls in compiled expression tree, and the provider recognizes the calls and translates to corresponding SQL.

Like functions listed [here](https://www.npgsql.org/efcore/mapping/translations.html), Npgsql can expose functions that maps to those.

**Reply by roji** [reply]
Indeed, this question doesn't belong in dotnet/runtime; it's a SQL translation question, and so belongs in EF, or even the specific EF provider(s) in question.
